src/v/kafka/security/scram_algorithm.cc

```diff
@@ -47,14 +60,98 @@ namespace kafka {
 // NOLINTNEXTLINE
 #define ALPHA "[a-zA-Z]"
 
-// clang-format off
-static constexpr auto client_first_message_re = ctll::fixed_string{
-    "n,(?:a=(" SASLNAME "))?,(m=" VALUE ",)?n=(" SASLNAME "),"
-    "r=(" PRINTABLE ")(," ALPHA "+=" VALUE ")*"};
+#define CLIENT_FIRST_MESSAGE_RE                                                \
+    "n,(?:a=(" SASLNAME "))?,(m=" VALUE ",)?n=(" SASLNAME "),"                 \
+    "r=(" PRINTABLE ")(," ALPHA "+=" VALUE ")*"
+
+#define CLIENT_FINAL_MESSAGE_RE                                                \
+    "c=(" BASE64 "),r=(" PRINTABLE ")" EXTENSIONS ",p=(" BASE64 ")"
 
+#ifdef USE_CTRE
+static constexpr auto client_first_message_re = ctll::fixed_string{
+  CLIENT_FIRST_MESSAGE_RE};
 static constexpr auto client_final_message_re = ctll::fixed_string{
-    "c=(" BASE64 "),r=(" PRINTABLE ")" EXTENSIONS ",p=(" BASE64 ")"};
-// clang-format on
+  CLIENT_FINAL_MESSAGE_RE};
+#else
+static const char* client_first_message_re = CLIENT_FIRST_MESSAGE_RE;
+static const char* client_final_message_re = CLIENT_FINAL_MESSAGE_RE;
+#endif
+
+struct client_first_match {
+    ss::sstring authzid;
+    ss::sstring username;
+    ss::sstring nonce;
+    ss::sstring extensions;
+};
+
+static inline std::optional<client_first_match>
+parse_client_first(std::string_view message) {
+#ifdef USE_CTRE
+    auto match = ctre::match<client_first_message_re>(message);
```

Comment:
this makes sense and is great work around. I'd make 2 additional methods. 
```
ctre_parse_client_first()
std_parse_client_first()
```
and then do the 
```
#ifdef USE_CTRE
return ctre_parse_client_first(message)
#else 
return std_parse_client_first(message)
```

so they can be tested independently. 

Would be good to make sure they pass the same unit test at all times tho. 
so you might want to forward declare them.

Reply:
good idea with the extra methods. that'll make the code easier to read.

but i'm not sure what your proposal is regarding changes to testing. the current set of tests (see scram_algorithm_test.cc) exercise this regex parsing. the difference being that clang uses ctre and gcc uses std::regex. do you have a different goal in mind?

Reply:
The idea was to always run scram-algorithm-test.cc with both impls

Reply:
isn't that what we are trying to do because we can't compile the regex's with ctre under gcc?

Reply:
good point. plus we run 4 builds (gcc, clang, debug, release) so it will get exercise w/ every build.